mne/preprocessing/maxwell.py

```diff
@@ -0,0 +1,565 @@
+# Authors: Mark Wronkiewicz <wronk.mark@gmail.com>
+#          Jussi Nurminen <jnu@iki.fi>
+
+# This code was adapted and relicensed (with BSD form) with permission from
+# Jussi Nurminen
+
+# License: BSD (3-clause)
+
+# Note, there are an absurd number of different notations for spherical
+# harmonics (partly because there is also no accepted standard for spherical
+# coordinates). Here, we purposefully stay away from shorthand notation in
+# both and use explicit terms (like 'azimuth' and 'polar') to avoid confusion.
+
+# TODO: write in equation numbers from Samu's paper
+
+import numpy as np
+from scipy.special import sph_harm, lpmv
+from scipy.linalg import pinv
+from ..forward._compute_forward import _concatenate_coils
+from ..forward._lead_dots import (_get_legen_table, _get_legen_lut_fast,
+                                  _get_legen_lut_accurate)
+#from ..fixes import partial
+from scipy.misc import factorial as fact
+
+
+def maxwell_filter(raw, coils, origin, int_order=8, ext_order=3, n_jobs=1):
+    """Apply Maxwell filter to data using spherical harmonics.
+
+    Parameters
+    ----------
+    raw : instance of mne.io.Raw
+        Data to be filtered
+    coils : list
+        List of MEG coils. Each element must a contain coil information dict
+        continaining 'rmag', 'cosmag', and 'w' keys
+    origin : tuple, shape (3,)
+        Origin of internal and external multipolar moment space
+    int_order : int
+        Order of internal component of spherical expansion
+    ext_order : int
+        Order of external component of spherical expansion
+    n_jobs : int
+        Number of jobs to run in parallel
+
+    Returns
+    -------
+    raw_sss : instance of mne.io.Raw
+        The raw data with Maxwell filtering applied
+    """
+
+    # TODO: Error checks on input parameters
+    picks = [raw.info['ch_names'].index(ch) for ch in [coil['chname']
+                                                       for coil in coils]]
+    data, times = raw[picks, :]
+
+    # Compute spherical harmonics
+    S_in, S_out = _sss_basis(origin, coils, int_order, ext_order)
+    S_tot = np.c_[S_in, S_out]
+
+    # Pseudo-inverse of total multipolar moment basis set
+    # TODO: determine appropriate pinv tolerance
+    pS_tot = pinv(S_tot, rcond=1e-15)
+    # Compute multipolar moments of data
+    mm = np.dot(pS_tot, data)
+    # Reconstruct data from internal space
+    recon = np.dot(S_in, mm[:80, :])
+
+    # Return raw file object
+    raw_sss = raw.copy()
+    raw_sss[:, :] = recon
```

Comment:
@jjnurminen @staulu, the above ~10 lines actually implement the Maxwell filter. Can you both check for any obvious errors?